### PR TITLE
Deprecate existing coupon codes methods

### PR DESCRIPTION
**Pull request by aitbw**
Hi! :wave:

Following @kennyadsl's idea on solidusio#2327, this PR aims to deprecate current endpoints used to apply coupon codes (on orders and checkout) and replace them with a new API endpoint properly named `Spree::Api::CouponCodesController`. This was also done for the `frontend` with `Spree::CouponCodesController`.

I think this approach allows a better separation of concerns and it's easier to reason about. Plus, it'd be easier to implement a built-in solution for solidusio#1641.

I'm open to suggestions!

### CODE REVIEW

File: api/app/controllers/spree/api/checkouts_controller.rb [hunk 1]
```diff
@@ -114,6 +114,7 @@ def state_callback(before_or_after = :before)
 
       def after_update_attributes
         if params[:order] && params[:order][:coupon_code].present?
+          Spree::Deprecation.warn('This method is deprecated. Please, use `Spree::Api::CouponCodesController` endpoints instead.')
```
[the remark this answers is not in the document]
Comment (aitbw) [reply api/app/controllers/spree/api/checkouts_controller.rb]: Hey @jacobherrington, thanks for your comments! I'll address them in a few. Just one question:

You mean a better message would include something like `Spree::Api::CouponCodesController#create`?

[the remark this answers is not in the document]
Comment (jacobherrington) [reply api/app/controllers/spree/api/checkouts_controller.rb]: I feel like I would rather be given a specific endpoint that is most similar to the method I attempted to use, does that make sense?

[the remark this answers is not in the document]
Comment (aitbw) [reply api/app/controllers/spree/api/checkouts_controller.rb]: Yeah, completely! I'll fix that too :ok_hand:

File: frontend/app/controllers/spree/checkout_controller.rb [hunk 2]
```diff
@@ -168,6 +168,7 @@ def completion_route
 
     def apply_coupon_code
       if update_params[:coupon_code].present?
+        Spree::Deprecation.warn('This endpoint is deprecated. Please, use `Spree::CouponCodesController` endpoints instead.')
```
[the remark this answers is not in the document]
Comment (aitbw) [reply frontend/app/controllers/spree/checkout_controller.rb]: Resolved via 4559a00

File: api/app/controllers/spree/api/orders_controller.rb [hunk 3]
```diff
@@ -92,6 +92,8 @@ def mine
       end
 
       def apply_coupon_code
+        Spree::Deprecation.warn('This endpoint is deprecated. Please, use `Spree::Api::CouponCodesController` endpoints instead.')
```
[the remark this answers is not in the document]
Comment (aitbw) [reply api/app/controllers/spree/api/orders_controller.rb]: Resolved via 4559a00

File: api/app/controllers/spree/api/checkouts_controller.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (aitbw) [reply api/app/controllers/spree/api/checkouts_controller.rb]: Resolved via 4559a00

**Comment by aitbw** [comment]
Noted @kennyadsl :ok_hand: let me have a look on that

File: api/spec/requests/spree/api/coupon_codes_controller_spec.rb [hunk 4]
```diff
@@ -12,6 +18,10 @@ module Spree
       let(:promo) { create(:promotion_with_item_adjustment, code: 'night_melody') }
       let(:promo_code) { promo.codes.first }
 
+      before do
+        allow_any_instance_of(Order).to receive_messages user: current_api_user
```
Comment (kennyadsl) [inline api/spec/requests/spree/api/coupon_codes_controller_spec.rb]: Does the test fail without this line? Also, maybe we can assign the user to the order in the factory instead, WDYT?

Comment (aitbw) [reply api/spec/requests/spree/api/coupon_codes_controller_spec.rb]: Yes, it fails **unless** we assign the user to the order, as you proposed. Both options are completely valid.

**Pull request merged**